src/coreclr/jit/morph.cpp

```diff
@@ -9672,15 +9672,19 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac, bool* optA
 #endif
             if (!varTypeIsFloating(tree->gtType))
             {
-                // We do not need to throw if the second operand is a non-(negative one) constant.
-                if (!op2->IsIntegralConst() || op2->IsIntegralConst(-1))
+                if (tree->CanDivisionPossiblyOverflow(this))
```

Comment:
It would be better to structure this in conjunction with `OperExceptions`. There are two ways to do that:
1) Call `CanDivisionPossiblyOverflow` in `OperExceptions` and call `OperExceptions` here. This is more expensive, but more consistent with the existing code and potentially more precise.
2) Let `OperExceptions` only check the new flags which are set here (i. e. delete the equivalent of `CanDivisionPossiblyOverflow` handling from it).

Reply:
I'm wondering if we should do this in a separate PR - I've already changed enough in this one to make the behavior better.

Reply:
Looking at OperException, appears we also duplicated some of the logic there too.

Reply:
Yep (hence my point). It is also duplicated in VN.

Reply:
So I will do it in this PR considering it's better to get rid of the duplicate logic.